src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/WhenShouldBeUsedCheck.java

```diff
@@ -0,0 +1,176 @@
+ * </p>
+ * <p>
+ * Parent is {@code com.puppycrawl.tools.checkstyle.TreeWalker}
+ * </p>
+ * <p>
+ * Violation Message Keys:
+ * </p>
+ * <ul>
+ * <li>
+ * {@code when.should.be.used}
+ * </li>
+ * </ul>
+ *
+ * @since 10.18.0
+ */
+
+@FileStatefulCheck
+public class WhenShouldBeUsedCheck extends AbstractCheck {
+
+    /**
+     * A key is pointing to the warning message text in "messages.properties"
+     * file.
+     */
+    public static final String MSG_KEY = "when.should.be.used";
+
+    /**
+     *  An array of acceptable statements inside the case block.
+     */
+    private static final int[] ACCEPTABLE_STATEMENTS_INSIDE_CASE_BLOCK = {
+        TokenTypes.LITERAL_IF,
+        TokenTypes.LITERAL_BREAK,
+        TokenTypes.EMPTY_STAT,
+        TokenTypes.RCURLY,
+    };
+
+    @Override
+    public int[] getDefaultTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] {TokenTypes.LITERAL_CASE};
+    }
+
+    @Override
+    public void visitToken(DetailAST ast) {
+        final DetailAST statementList = getStatementListOfCase(ast);
+        final boolean hasPatternLabel = hasPatternLabel(ast);
+        if (hasPatternLabel) {
+            final List<DetailAST> blockStatements = getBlockStatements(statementList);
+            final boolean hasAcceptableStatementsOnly = blockStatements.stream()
+                .allMatch(WhenShouldBeUsedCheck::isAcceptableStatement);
+            final boolean hasSingleIfWithNoElse = blockStatements.stream()
+                .filter(statement -> {
+                    return statement.getType() == TokenTypes.LITERAL_IF
+                            && statement.findFirstToken(TokenTypes.LITERAL_ELSE) == null;
+                }).count() == 1;
+            if (hasAcceptableStatementsOnly && hasSingleIfWithNoElse) {
+                log(ast, MSG_KEY);
+            }
+        }
+    }
+
+    /**
+     * Get the statement list token of the case block.
+     *
+     * @param caseAST the AST node representing {@code LITERAL_CASE}
+     * @return the AST node representing {@code SLIST} of the current case
+     */
+    private static DetailAST getStatementListOfCase(DetailAST caseAST) {
+        final DetailAST caseParent = caseAST.getParent();
+        final DetailAST statementList = caseParent.findFirstToken(TokenTypes.SLIST);
+
+        return Optional.ofNullable(statementList)
+               .map(ast -> ast.findFirstToken(TokenTypes.SLIST))
+               .orElse(statementList);
+    }
+
+    /**
+     * Get all statements inside the case block.
+     *
+     * @param statementList the AST node representing {@code SLIST} of the current case
+     * @return statements inside the current case block
+     */
+    private static List<DetailAST> getBlockStatements(DetailAST statementList) {
```

Comment:
Ideally, we should never pass in a null DetailAST here. How can we prevent this?

Reply:
I just added another condition `hasPatternLabel && statementList != null` if one of them is false we don't need to continue